### ISSUE TITLE
Improve numpy.interp running time (migrated from Trac #1920)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3468]
Original ticket http://projects.scipy.org/numpy/ticket/1920
Reported 2011-08-01 by trac user tkluck, assigned to unknown.
## (copy-pasted from numpy-discussion, but extra comment below)

The current implementation of numpy.interp(x,xp,fp) comes down to: first calculating all the slopes of the linear interpolant (these are len(xp)-1), then use a binary search to find where x is in xp (running time log(len(xp)). So we obtain a running time of

O( len(xp) + len(x)*log(len(xp) )

We could improve this to just

O( len(x)*log(len(xp) )

by not caching the slopes. The point is, of course, that this is slightly slower in the common use case where x is is refinement of xp, and where you will have to compute all the slopes anyway.
## In my personal use case, however, I needed the value of the interp(x0,xp,fp) in order to calculate the next point x1 where I wanted to calculate interp(x1,xp,fp). The current implementation gave a severe running time penalty.

After a short discussion on the mailinglist, it was thought to be a good idea to offer both implementations: to cache the slopes when len(xp) <= len(x), and to not do so otherwise. In the first cases caching gives about 10% increase in speed (in my not-so-thorough testing), which I feel justifies the (minor) extra code complexity.

Patch attached.

**Comment by thouis** [comment]
Attachment in Trac by trac user tkluck, 2011-08-01: [0001-improved-numpy.interp-running-time.patch](http://projects.scipy.org/numpy/attachment/ticket/1920/0001-improved-numpy.interp-running-time.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-16

I'm not 100% sure I understand the statements on speed.  You have two cases:
1. len(xp) <= len(x)
2. len(xp) > len(x)

The new code is almost unchanged for case 1, and is "slightly slower".
For case 2, caching the slopes is removed, resulting in a 10% speed-up. Correct or not? If so, I'm not sure I'd call that a "severe speed penalty". Can you provide a snippet used to measure the speed difference?

Looking at the tests, only case 1 is tested. So your patch would need some tests. Also a few more comments in the code about that the if/else is there for speed reasons etc. would be helpful.

**Comment by thouis** [comment]
Comment in Trac by trac user tkluck, 2011-08-16

Replying to [comment:1 rgommers]:

> I'm not 100% sure I understand the statements on speed.  You have two cases:
> 1. len(xp) <= len(x)
> 2. len(xp) > len(x)
> 
> The new code is almost unchanged for case 1, and is "slightly slower".
> For case 2, caching the slopes is removed, resulting in a 10% speed-up. Correct or not? If so, I'm not sure I'd call that a "severe speed penalty". Can you provide a snippet used to measure the speed difference?

Thanks for commenting.

The 'severe running time penalty' refers to case 2: pre-calculating the slopes requires a loop that is executed len(xp) times. When this is large (as compared to len(x)), there is no need to calculate that many slopes -- it can happen that you calculate 40,000 slopes just to find the interpolation at a single point x.

So, in terms of asymptotical running time, it is optimal _not_ to pre-calculate any slopes. My first idea, therefore, was to do away with pre-calculating the slopes entirely. This results in optimal asymptotical running time.

However, it is probably very common that x is a refinement of xp. In that case, pre-caching the slopes does make a difference -- effectively, you're replacing a floating point calculation by a memory lookup. I measured a 10% speed increase (I'll try to find my tests again and post them here). Therefore, I decided to introduce the branching depending on which case we're in.

> Looking at the tests, only case 1 is tested. So your patch would need some tests. Also a few more comments in the code about that the if/else is there for speed reasons etc. would be helpful.
> I'm not familiar yet with the test suite. I'll try to look into that and update my patch accordingly.

**Comment by thouis** [comment]
Attachment in Trac by trac user tkluck, 2011-08-18: [numpy_interp_bench.c](http://projects.scipy.org/numpy/attachment/ticket/1920/numpy_interp_bench.c)

**Comment by thouis** [comment]
Comment in Trac by trac user tkluck, 2011-08-18

Here's my very simple comparison of pre-calculating the slopes and not doing so. I attached a program in C {{{numpy_interp_bench.c}}} that compares the calculation with and without precalculation. The results are:

```
tkluckatmention:notebook-tjk:~$ time ./numpy_interp_bench 100000000 10 slopes
interpolating a function with 10 control points at 100000000 points, with slope cache

real    0m11.277s
user    0m9.940s
sys 0m1.150s
tkluckatmention:notebook-tjk:~$ time ./numpy_interp_bench 100000000 10 noslopes
interpolating a function with 10 control points at 100000000 points, without slope cache

real    0m12.292s
user    0m11.020s
sys 0m1.140s
```

This was what I meant by my statement that in the case that len(x) > len(xp), it makes a ~ 10% difference to pre-calculate the slopes.

**Comment by thouis** [comment]
Attachment in Trac by trac user tkluck, 2011-08-22: [0002-Added-doctests-for-np.interp-for-the-case-where-len-.patch](http://projects.scipy.org/numpy/attachment/ticket/1920/0002-Added-doctests-for-np.interp-for-the-case-where-len-.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user tkluck, 2011-08-22

I just added a patch that adds doctesting for the len(x) < len(xp) codepath.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-27

A few comments on testing:

```
- doctests are not part of the test suite (not run by default), they're for examples.
- instead of rounding a float, the function to use is {{{assert_almost_equal}}} from numpy.testing
```

Could you please update your test so it integrates with the testing framework? Guidelines are at https://github.com/numpy/numpy/blob/master/doc/TESTS.rst.txt, please ask if you need more help.

**Comment by thouis** [comment]
Attachment in Trac by trac user tkluck, 2011-10-18: [0003-Revert-Added-doctests-for-np.interp-for-the-case-whe.patch](http://projects.scipy.org/numpy/attachment/ticket/1920/0003-Revert-Added-doctests-for-np.interp-for-the-case-whe.patch)

**Comment by thouis** [comment]
Attachment in Trac by trac user tkluck, 2011-10-18: [0004-Added-a-test-in-the-Nose-test-framework.patch](http://projects.scipy.org/numpy/attachment/ticket/1920/0004-Added-a-test-in-the-Nose-test-framework.patch)

**Comment by thouis** [comment]
Comment in Trac by trac user tkluck, 2011-10-18

Thanks for your comments, and sorry for taking so long to implement them. I moved the tests to the Nose framework in patches 0003 and 0004. Please review.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-04-01

Fixed up a bit and pushed in b9576ed.
